src/Chapter3.hs

```diff
@@ -479,6 +521,35 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+data House = House {housePeopleNum::Int} deriving (Show)
+data MainBuilding = Church | Library deriving (Show)
+data Castle = Castle {castleName::String} | NoCastle deriving (Show)
+data City = City
+  { cityCastle :: Castle
+  , cityHasWall :: Bool
+  , cityMain :: MainBuilding
+  , cityHouses :: [House]
+  } deriving (Show)
+
+buildCastle :: String -> City -> City
+buildCastle name city = city {cityCastle=Castle name}
+
+buildHouse :: Int -> City -> City
+buildHouse peopleNum city 
+  | peopleNum>=1 && peopleNum<=4 = city {cityHouses=House peopleNum : cityHouses city}
```

Comment:
Interesting approach. It is usually called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `mkHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂

Reply:
@timqsh For this particular task it's good enough because the allowed number of people if not big :slightly_smiling_face: 
If the requirements will be something like between 1 and 1000, you can have a newtype around `Int` and use a smart constructor pattern :slightly_smiling_face: